src/System.Net.Http/src/System/Net/Http/Managed/HttpConnectionPool.cs

```diff
@@ -2,130 +2,440 @@
+                // to complicate the logic about trying to get a different pool when the
+                // retrieved one has been disposed of.  In the future we could alternatively
+                // try returning such connections to whatever pool is currently considered
+                // current for that endpoint, if there is one.
             }
-
-            return null;
         }
 
-        public void AddConnection(HttpConnection connection)
+        /// <summary>Waits for and returns the created connection, decrementing the associated connection count if it fails.</summary>
+        private async ValueTask<HttpConnection> WaitForCreatedConnectionAsync(ValueTask<HttpConnection> creationTask)
         {
-            if (!_activeConnections.TryAdd(connection))
+            try
             {
-                throw new InvalidOperationException();
+                return await creationTask.ConfigureAwait(false);
+            }
+            catch
+            {
+                DecrementConnectionCount();
+                throw;
             }
         }
 
-        public void PutConnection(HttpConnection connection)
+        /// <summary>
+        /// Increments the count of connections associated with the pool.  This is invoked
+        /// any time a new connection is created for the pool.
+        /// </summary>
+        public void IncrementConnectionCount()
         {
-            if (!_activeConnections.TryRemove(connection))
+            lock (SyncObj) IncrementConnectionCountNoLock();
+        }
+
+        private void IncrementConnectionCountNoLock()
+        {
+            Debug.Assert(Monitor.IsEntered(SyncObj), $"Expected to be holding {nameof(SyncObj)}");
+
+            _usedSinceLastCleanup = true;
+
+            if (_associatedConnectionCount == NewPoolSentinel)
             {
-                throw new InvalidOperationException();
+                _associatedConnectionCount = 1;
+            }
+            else
+            {
+                Debug.Assert(
+                    _associatedConnectionCount >= 0 && _associatedConnectionCount < _maxConnections,
+                    $"Expected 0 <= {_associatedConnectionCount} < {_maxConnections}");
+                _associatedConnectionCount++;
             }
-
-            _idleConnections.Add(connection);
         }
 
-        public void Dispose()
+        /// <summary>
+        /// Decrements the number of connections associated with the pool.
+        /// If there are waiters on the pool due to having reached the maximum,
+        /// this will instead try to transfer the count to one of them.
+        /// </summary>
+        public void DecrementConnectionCount()
         {
-            if (!_disposed)
+            lock (SyncObj)
             {
-                _disposed = true;
+                Debug.Assert(_associatedConnectionCount > 0, $"Expected {_associatedConnectionCount} > 0");
```

Comment:
Didn't seem to be necessary given that we're decrementing here rather than incrementing.  But I can add it if you think it's valuable.